dev/TeachingTip/TeachingTip.cpp

```diff
@@ -1208,249 +1255,391 @@ void TeachingTip::StartContractToClose()
+    {
+        winrt::Thickness screenSpaceAroundTarget{ targetBounds.X + windowBounds.X,
+                                                  targetBounds.Y + windowBounds.Y,
+                                                  screenBounds.Width - (windowBounds.X + targetBounds.X + targetBounds.Width),
+                                                  screenBounds.Height - (windowBounds.Y + targetBounds.Y + targetBounds.Height) };
+
+        return std::tuple<winrt::Rect, winrt::Thickness, winrt::Thickness>(targetBounds, windowSpaceAroundTarget, screenSpaceAroundTarget);
+    }
+    else
+    {
+        return std::tuple<winrt::Rect, winrt::Thickness, winrt::Thickness>(targetBounds, windowSpaceAroundTarget, windowSpaceAroundTarget);
+    }
+}
+
+std::list<winrt::TeachingTipPlacementMode> TeachingTip::GetPlacementFallbackOrder(winrt::TeachingTipPlacementMode preferredPlacement)
+{
+    auto priority1 = winrt::TeachingTipPlacementMode::Top;
+    auto priority2 = winrt::TeachingTipPlacementMode::Bottom;
+    auto priority3 = winrt::TeachingTipPlacementMode::Left;
+    auto priority4 = winrt::TeachingTipPlacementMode::Right;
+    auto priority5 = winrt::TeachingTipPlacementMode::TopEdgeAlignedLeft;
+    auto priority6 = winrt::TeachingTipPlacementMode::TopEdgeAlignedRight;
+    auto priority7 = winrt::TeachingTipPlacementMode::BottomEdgeAlignedLeft;
+    auto priority8 = winrt::TeachingTipPlacementMode::BottomEdgeAlignedRight;
+    auto priority9 = winrt::TeachingTipPlacementMode::LeftEdgeAlignedTop;
+    auto priority10 = winrt::TeachingTipPlacementMode::LeftEdgeAlignedBottom;
+    auto priority11 = winrt::TeachingTipPlacementMode::RightEdgeAlignedTop;
+    auto priority12 = winrt::TeachingTipPlacementMode::RightEdgeAlignedBottom;
+
+
+    if (isPlacementBottom(preferredPlacement))
+    {
+        // Swap to bottom > top
+        std::swap(priority1, priority2);
+        std::swap(priority5, priority7);
+        std::swap(priority6, priority8);
+    }
+    else if (isPlacementLeft(preferredPlacement))
+    {
+        // swap to lateral > vertical
+        std::swap(priority1, priority3);
+        std::swap(priority2, priority4);
+        std::swap(priority5, priority9);
+        std::swap(priority6, priority10);
+        std::swap(priority7, priority11);
+        std::swap(priority8, priority12);
+    }
+    else if (isPlacementRight(preferredPlacement))
+    {
+        // swap to lateral > vertical
+        std::swap(priority1, priority3);
+        std::swap(priority2, priority4);
+        std::swap(priority5, priority9);
+        std::swap(priority6, priority10);
+        std::swap(priority7, priority11);
+        std::swap(priority8, priority12);
+
+        // swap to right > left
+        std::swap(priority1, priority2);
+        std::swap(priority5, priority7);
+        std::swap(priority6, priority8);
+    }
+
+    auto priorityList = std::list<winrt::TeachingTipPlacementMode>();
+    priorityList.push_back(priority1);
```

Comment:
prefer std::vector. List is pretty inefficient for what we do (separate allocation for each element).

Reply:
Looking at the code more, you know the length since it's always 12 positions. Recommend using std::array<winrt::TeachingTipPlacementMode, 12>.

---
In reply to: [269287616](https://github.com/Microsoft/microsoft-ui-xaml/pull/480#discussion_r269287616) [](ancestors = 269287616)